File: client/src/types/ranking.ts
import { CountryModel } from './country'

export type Ranking = {
  id: number
  name: string
  name2: string
  isTeam: boolean
  categories: RankingCategory[]
}

export type RankingData = {
  publication_date: string // Date
}

export type RankingCategory = {
  id: number
  name: string
  shortName: string
  isDoubles: boolean
}

export type RankingWeek = {
  date: string // Date
  display: string
  id: number
  key: string
  week: number
  year: number
}

export type RankingSearch = {
  rankingId: number
  categoryId: number
  weekId?: number
  isDoubles?: boolean
  page?: number
  perPage?: number
}

export type RankingSearchResponse = {
  entries: PlayerRankingEntry[] | TeamRankingEntry[]
  total: number
  lastPage: number
  currentPage: number
  onThisPage: number
}

export interface RankingEntry {
  confederation_id: number
  id: number
  rank: number
  rank_change: number
  ranking_category_id: number
}

export interface PlayerRankingEntry extends RankingEntry {
  p1_country: string
  player1_id: number
  player1_model: PlayerModel
  p2_country: string | null
  player2_id: number | null
  player2_model: PlayerModel | null
  points: string
  rank_previous: number
  tournaments: number
}

export interface TeamRankingEntry extends RankingEntry {
  team_id: number
  team_ms: string
  team_ws: string
  team_md: string
  team_wd: string
  team_xd: string
  team_sc: string
  team_tc: string
  team_uc: string
  team_total_points: string
  team_model: TeamModel
}

export type TeamModel = {
  code: string
  country_code: string
  country_id: number
  country_model: CountryModel
  name: string
  type: number
}

export type PlayerModel = {
  active: number
  code: string
  country: string
  country_model: CountryModel
  date_of_birth: string
  first_name: string
  gender_id: number
  id: number
  last_name: string
  name_display: string
  nationality: string
  slug: string
}

export type PlayerPointsBreakdown = {
  date: string
  name: string
  points: string
  
  ranking_calc: 1 | 0
  url: string

  result: string
  tournamentId: number
  isCountedIn: boolean
}

File: server/src/transformers/ranking.py
from typing import Dict, Any, List
from utils.country import add_flag_url_to_country
import re


def transform_ranking_table_response(response: dict):
    cleaned_data: List[Dict[str, Any]] = []
    results = response.get("results")

    for ranking_entry in results.get("data"):
        player1 = ranking_entry.get("player1_model")
        player2 = ranking_entry.get("player2_model")
        team = ranking_entry.get("team_model")

        if player1 is not None:
            add_flag_url_to_country(player1.get("country_model"))

        if player2 is not None:
            add_flag_url_to_country(player2.get("country_model"))

        if team is not None:
            add_flag_url_to_country(team.get("country_model"))

        cleaned_data.append(ranking_entry)

    results_from = results.get("from")
    results_to = results.get("to")

    on_this_page = 0
    if results_from and results_to:
        on_this_page = results_to - results_from + 1

    return {
        "lastPage": results.get("last_page"),
        "currentPage": results.get("current_page"),
        "total": results.get("total"),
        "entries": cleaned_data,
        "onThisPage": on_this_page,
    }


def transform_ranking_data_response(response: dict):
    return response[0]


def transform_breakdown_response(response: List):
    cleaned_data: List[Dict[str, Any]] = []

    for tournament in response:
        tournament_id = re.search(r"/tournament/(\d+)/", tournament.get("url"))
        tournament["tournamentId"] = tournament_id.group(1) if tournament_id else None

        ranking_calc = tournament.get("ranking_calc", 0)
        tournament["isCountedIn"] = True if ranking_calc == 1 else False

        result = tournament.get("result")
        tournament["result"] = f"{result}/{result * 2 - 2}" if result > 2 else result 

        cleaned_data.append(tournament)

    return cleaned_data
